TrainBertOnCode demo does not work

## Description

Running the [`TrainBertOnCode`](https://github.com/awslabs/djl/blob/master/examples/src/main/java/ai/djl/examples/training/TrainBertOnCode.java) demo fails with an error.

### Expected Behavior

`TrainBertOnCode` should not fail.

### Error Message

```
> Task :examples:run
[INFO ] - Training on: 1 GPUs.
[INFO ] - Load MXNet Engine Version 1.7.0 in 0.067 ms.
[INFO ] - Training on: 1 GPUs.
[INFO ] - Load MXNet Engine Version 1.7.0 in 0.020 ms.
[ERROR] - Temporary resource return failed.
java.lang.NullPointerException: null
        at ai.djl.mxnet.engine.MxNDArray.detach(MxNDArray.java:184) ~[mxnet-engine-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.mxnet.engine.MxNDArray.attach(MxNDArray.java:167) ~[mxnet-engine-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.ndarray.BaseNDManager.close(BaseNDManager.java:262) [api-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.nn.transformer.BertPretrainingBlock.forwardInternal(BertPretrainingBlock.java:73) [api-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.nn.AbstractBlock.forwardInternal(AbstractBlock.java:170) [api-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.nn.AbstractBlock.forward(AbstractBlock.java:135) [api-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.training.Trainer.forward(Trainer.java:173) [api-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.training.EasyTrain.trainSplit(EasyTrain.java:120) [api-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.training.EasyTrain.trainBatch(EasyTrain.java:108) [api-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.examples.training.TrainBertOnCode.runExample(TrainBertOnCode.java:120) [main/:?]
        at ai.djl.examples.training.TrainBertOnCode.main(TrainBertOnCode.java:86) [main/:?]
[ERROR] - Temporary resource return failed.
java.lang.NullPointerException: null
        at ai.djl.mxnet.engine.MxNDArray.detach(MxNDArray.java:184) ~[mxnet-engine-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.mxnet.engine.MxNDArray.attach(MxNDArray.java:167) ~[mxnet-engine-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.ndarray.BaseNDManager.close(BaseNDManager.java:262) [api-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.nn.transformer.BertPretrainingBlock.forwardInternal(BertPretrainingBlock.java:73) [api-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.nn.AbstractBlock.forwardInternal(AbstractBlock.java:170) [api-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.nn.AbstractBlock.forward(AbstractBlock.java:135) [api-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.training.Trainer.forward(Trainer.java:173) [api-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.training.EasyTrain.trainSplit(EasyTrain.java:120) [api-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.training.EasyTrain.trainBatch(EasyTrain.java:108) [api-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.examples.training.TrainBertOnCode.runExample(TrainBertOnCode.java:120) [main/:?]
        at ai.djl.examples.training.TrainBertOnCode.main(TrainBertOnCode.java:86) [main/:?]
[ERROR] - Temporary resource return failed.
java.lang.NullPointerException: null
        at ai.djl.mxnet.engine.MxNDArray.detach(MxNDArray.java:184) ~[mxnet-engine-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.mxnet.engine.MxNDArray.attach(MxNDArray.java:167) ~[mxnet-engine-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.ndarray.BaseNDManager.close(BaseNDManager.java:262) [api-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.nn.transformer.BertPretrainingBlock.forwardInternal(BertPretrainingBlock.java:73) [api-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.nn.AbstractBlock.forwardInternal(AbstractBlock.java:170) [api-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.nn.AbstractBlock.forward(AbstractBlock.java:135) [api-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.training.Trainer.forward(Trainer.java:173) [api-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.training.EasyTrain.trainSplit(EasyTrain.java:120) [api-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.training.EasyTrain.trainBatch(EasyTrain.java:108) [api-0.11.0-SNAPSHOT.jar:?]
        at ai.djl.examples.training.TrainBertOnCode.runExample(TrainBertOnCode.java:120) [main/:?]
        at ai.djl.examples.training.TrainBertOnCode.main(TrainBertOnCode.java:86) [main/:?]
Exception in thread "main" ai.djl.engine.EngineException: MXNet engine call failed: MXNetError: Shape inconsistent, Provided = [512,512], inferred shape=(512,2048)

        at ai.djl.mxnet.jna.JnaUtils.checkCall(JnaUtils.java:1930)
        at ai.djl.mxnet.jna.JnaUtils.imperativeInvoke(JnaUtils.java:519)
        at ai.djl.mxnet.jna.FunctionInfo.invoke(FunctionInfo.java:72)
        at ai.djl.mxnet.engine.MxNDManager.invoke(MxNDManager.java:288)
        at ai.djl.mxnet.engine.MxNDArrayEx.linear(MxNDArrayEx.java:599)
        at ai.djl.nn.core.Linear.linear(Linear.java:178)
        at ai.djl.nn.core.Linear.forwardInternal(Linear.java:87)
        at ai.djl.nn.AbstractBlock.forward(AbstractBlock.java:121)
        at ai.djl.nn.Block.forward(Block.java:122)
        at ai.djl.nn.transformer.PointwiseFeedForwardBlock.forwardInternal(PointwiseFeedForwardBlock.java:98)
        at ai.djl.nn.AbstractBlock.forward(AbstractBlock.java:121)
        at ai.djl.nn.Block.forward(Block.java:122)
        at ai.djl.nn.transformer.TransformerEncoderBlock.forwardInternal(TransformerEncoderBlock.java:113)
        at ai.djl.nn.AbstractBlock.forward(AbstractBlock.java:121)
        at ai.djl.nn.Block.forward(Block.java:122)
        at ai.djl.nn.transformer.BertBlock.forwardInternal(BertBlock.java:252)
        at ai.djl.nn.AbstractBlock.forward(AbstractBlock.java:121)
        at ai.djl.nn.Block.forward(Block.java:122)
        at ai.djl.nn.transformer.BertPretrainingBlock.forwardInternal(BertPretrainingBlock.java:77)
        at ai.djl.nn.AbstractBlock.forwardInternal(AbstractBlock.java:170)
        at ai.djl.nn.AbstractBlock.forward(AbstractBlock.java:135)
        at ai.djl.training.Trainer.forward(Trainer.java:173)
        at ai.djl.training.EasyTrain.trainSplit(EasyTrain.java:120)
        at ai.djl.training.EasyTrain.trainBatch(EasyTrain.java:108)
        at ai.djl.examples.training.TrainBertOnCode.runExample(TrainBertOnCode.java:120)
        at ai.djl.examples.training.TrainBertOnCode.main(TrainBertOnCode.java:86)

> Task :examples:run FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':examples:run'.
> Process 'command '/usr/lib/jvm/jdk-11.0.9+11/bin/java'' finished with non-zero exit value 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.7.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 6s
23 actionable tasks: 1 executed, 22 up-to-date
```

## How to Reproduce?

Run the `TrainBertOnCode` demo.

### Steps to reproduce
(Paste the commands you ran that produced the error.)

1. `cd examples`
2. `./gradlew run -Dmain=ai.djl.examples.training.TrainBertOnCode`

## What have you tried to solve it?

1. Changing [`maxSequenceLength`](https://github.com/awslabs/djl/blob/fc1290f48f1d55592f710f8a6b20b4c88ae2f6f7/examples/src/main/java/ai/djl/examples/training/TrainBertOnCode.java#L107) to `2048`.

## Environment Info

Please run the command `./gradlew debugEnv` from the root directory of DJL (if necessary, clone DJL first). It will output information about your system, environment, and installation that can help us debug your issue. Paste the output of the command below:

```

> Task :pytorch:pytorch-engine:processResources
Downloading https://publish.djl.ai/pytorch-1.7.1/jnilib/0.11.0/linux-x86_64/cpu/libdjl_torch.so
Downloading https://publish.djl.ai/pytorch-1.7.1/jnilib/0.11.0/linux-x86_64/cu101/libdjl_torch.so
Downloading https://publish.djl.ai/pytorch-1.7.1/jnilib/0.11.0/linux-x86_64/cu102/libdjl_torch.so
Downloading https://publish.djl.ai/pytorch-1.7.1/jnilib/0.11.0/linux-x86_64/cu110/libdjl_torch.so
Downloading https://publish.djl.ai/pytorch-1.7.1/jnilib/precxx11/0.11.0/linux-x86_64/cpu/libdjl_torch.so
Downloading https://publish.djl.ai/pytorch-1.7.1/jnilib/0.11.0/osx-x86_64/cpu/libdjl_torch.dylib
Downloading https://publish.djl.ai/pytorch-1.7.1/jnilib/0.11.0/win-x86_64/cpu/djl_torch.dll
Downloading https://publish.djl.ai/pytorch-1.7.1/jnilib/0.11.0/win-x86_64/cu101/djl_torch.dll
Downloading https://publish.djl.ai/pytorch-1.7.1/jnilib/0.11.0/win-x86_64/cu102/djl_torch.dll
Downloading https://publish.djl.ai/pytorch-1.7.1/jnilib/0.11.0/win-x86_64/cu110/djl_torch.dll

> Task :integration:debugEnv
[DEBUG] - Using cache dir: /home/breandan/.djl.ai/mxnet
[DEBUG] - Loading mxnet library from: /home/breandan/.djl.ai/mxnet/1.7.0-backport-cu102mkl-linux-x86_64/libmxnet.so
[DEBUG] - Engine loaded from provider: MXNet
[DEBUG] - Using cache dir: /home/breandan/.djl.ai/pytorch
[INFO ] - Downloading https://publish.djl.ai/pytorch-1.7.1/cu102/linux/native/lib/libc10_cuda.so.gz ...
[INFO ] - Downloading https://publish.djl.ai/pytorch-1.7.1/cu102/linux/native/lib/libc10.so.gz ...
[INFO ] - Downloading https://publish.djl.ai/pytorch-1.7.1/cu102/linux/native/lib/libcaffe2_nvrtc.so.gz ...
[INFO ] - Downloading https://publish.djl.ai/pytorch-1.7.1/cu102/linux/native/lib/libcudart-80664282.so.10.2.gz ...
[INFO ] - Downloading https://publish.djl.ai/pytorch-1.7.1/cu102/linux/native/lib/libgomp-75eea7e8.so.1.gz ...
[INFO ] - Downloading https://publish.djl.ai/pytorch-1.7.1/cu102/linux/native/lib/libnvrtc-08c4863f.so.10.2.gz ...
[INFO ] - Downloading https://publish.djl.ai/pytorch-1.7.1/cu102/linux/native/lib/libnvToolsExt-3965bdd0.so.1.gz ...
[INFO ] - Downloading https://publish.djl.ai/pytorch-1.7.1/cu102/linux/native/lib/libtorch_cpu.so.gz ...
[INFO ] - Downloading https://publish.djl.ai/pytorch-1.7.1/cu102/linux/native/lib/libtorch_cuda.so.gz ...
[INFO ] - Downloading https://publish.djl.ai/pytorch-1.7.1/cu102/linux/native/lib/libtorch.so.gz ...
[DEBUG] - Loading pytorch library from: /home/breandan/.djl.ai/pytorch/1.7.1-cu102-linux-x86_64/0.11.0-SNAPSHOT-cu102-libdjl_torch.so
[INFO ] - Number of inter-op threads is 4
[INFO ] - Number of intra-op threads is 8
[DEBUG] - Engine loaded from provider: PyTorch
[DEBUG] - Using cache dir: /home/breandan/.djl.ai/tensorflow
[INFO ] - Downloading https://publish.djl.ai/tensorflow-2.4.1/linux/cpu/libtensorflow_cc.so.2.gz ...
[INFO ] - Downloading https://publish.djl.ai/tensorflow-2.4.1/linux/cpu/THIRD_PARTY_TF_JNI_LICENSES.gz ...
[INFO ] - Downloading https://publish.djl.ai/tensorflow-2.4.1/linux/cpu/libtensorflow_framework.so.2.gz ...
[INFO ] - Downloading https://publish.djl.ai/tensorflow-2.4.1/linux/cpu/LICENSE.gz ...
[INFO ] - Downloading https://publish.djl.ai/tensorflow-2.4.1/linux/cpu/libjnitensorflow.so.gz ...
[DEBUG] - Loading TensorFlow library from: /home/breandan/.djl.ai/tensorflow/2.4.1-SNAPSHOT-20210317-cpu-linux-x86_64/libjnitensorflow.so
Warning: Could not load Loader: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path: [/usr/local/cuda-10.2/lib64, /usr/java/packages/lib, /usr/lib64, /lib64, /lib, /usr/lib]
Warning: Could not load Pointer: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path: [/usr/local/cuda-10.2/lib64, /usr/java/packages/lib, /usr/lib64, /lib64, /lib, /usr/lib]
Warning: Could not load BytePointer: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path: [/usr/local/cuda-10.2/lib64, /usr/java/packages/lib, /usr/lib64, /lib64, /lib, /usr/lib]
2021-03-20 22:32:06.549653: I external/org_tensorflow/tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
[DEBUG] - Engine loaded from provider: TensorFlow
[WARN ] - More than one deep learning engines found.
[DEBUG] - Found default engine: MXNet
----------- System Properties -----------
sun.desktop: gnome
awt.toolkit: sun.awt.X11.XToolkit
java.specification.version: 11
sun.cpu.isalist: 
sun.jnu.encoding: UTF-8
java.class.path: /home/breandan/IdeaProjects/djl/integration/build/classes/java/main:/home/breandan/IdeaProjects/djl/integration/build/resources/main:/home/breandan/.gradle/caches/modules-2/files-2.1/commons-cli/commons-cli/1.4/c51c00206bb913cd8612b24abd9fa98ae89719b1/commons-cli-1.4.jar:/home/breandan/.gradle/caches/modules-2/files-2.1/org.apache.logging.log4j/log4j-slf4j-impl/2.13.3/7cca27a921a18645139cf651c04b83b1a19cfd76/log4j-slf4j-impl-2.13.3.jar:/home/breandan/IdeaProjects/djl/basicdataset/build/libs/basicdataset-0.11.0-SNAPSHOT.jar:/home/breandan/IdeaProjects/djl/model-zoo/build/libs/model-zoo-0.11.0-SNAPSHOT.jar:/home/breandan/IdeaProjects/djl/testing/build/libs/testing-0.11.0-SNAPSHOT.jar:/home/breandan/.gradle/caches/modules-2/files-2.1/org.testng/testng/7.1.0/b0bcea778fb2899aeb4014c558babea8833d180a/testng-7.1.0.jar:/home/breandan/IdeaProjects/djl/mxnet/mxnet-model-zoo/build/libs/mxnet-model-zoo-0.11.0-SNAPSHOT.jar:/home/breandan/.gradle/caches/modules-2/files-2.1/ai.djl.mxnet/mxnet-native-auto/1.7.0-backport/ee5b368ef94c1fcec4ade4a6edacffb420fefce7/mxnet-native-auto-1.7.0-backport.jar:/home/breandan/IdeaProjects/djl/pytorch/pytorch-model-zoo/build/libs/pytorch-model-zoo-0.11.0-SNAPSHOT.jar:/home/breandan/.gradle/caches/modules-2/files-2.1/ai.djl.pytorch/pytorch-native-auto/1.7.1/5bae849c38d8bfa4c7a6963f7b6461e45157ee87/pytorch-native-auto-1.7.1.jar:/home/breandan/IdeaProjects/djl/tensorflow/tensorflow-model-zoo/build/libs/tensorflow-model-zoo-0.11.0-SNAPSHOT.jar:/home/breandan/.gradle/caches/modules-2/files-2.1/ai.djl.tensorflow/tensorflow-native-auto/2.4.1-SNAPSHOT/3ab29a57a3adfaf8292f3297579a09ec7be2a0ec/tensorflow-native-auto-2.4.1-SNAPSHOT.jar:/home/breandan/IdeaProjects/djl/mxnet/mxnet-engine/build/libs/mxnet-engine-0.11.0-SNAPSHOT.jar:/home/breandan/IdeaProjects/djl/pytorch/pytorch-engine/build/libs/pytorch-engine-0.11.0-SNAPSHOT.jar:/home/breandan/IdeaProjects/djl/tensorflow/tensorflow-engine/build/libs/tensorflow-engine-0.11.0-SNAPSHOT.jar:/home/breandan/IdeaProjects/djl/api/build/libs/api-0.11.0-SNAPSHOT.jar:/home/breandan/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-api/1.7.30/b5a4b6d16ab13e34a88fae84c35cd5d68cac922c/slf4j-api-1.7.30.jar:/home/breandan/.gradle/caches/modules-2/files-2.1/org.apache.logging.log4j/log4j-core/2.13.3/4e857439fc4fe974d212adaaaa3b118b8b50e3ec/log4j-core-2.13.3.jar:/home/breandan/.gradle/caches/modules-2/files-2.1/org.apache.logging.log4j/log4j-api/2.13.3/ec1508160b93d274b1add34419b897bae84c6ca9/log4j-api-2.13.3.jar:/home/breandan/.gradle/caches/modules-2/files-2.1/org.apache.commons/commons-csv/1.8/37ca9a9aa2d4be2599e55506a6d3170dd7a3df4/commons-csv-1.8.jar:/home/breandan/.gradle/caches/modules-2/files-2.1/com.beust/jcommander/1.72/6375e521c1e11d6563d4f25a07ce124ccf8cd171/jcommander-1.72.jar:/home/breandan/.gradle/caches/modules-2/files-2.1/com.google.inject/guice/4.1.0/faf9ee8ac09eafd1128091426dd367a8c0085d55/guice-4.1.0-no_aop.jar:/home/breandan/.gradle/caches/modules-2/files-2.1/org.yaml/snakeyaml/1.21/18775fdda48574784f40b47bf478ab0593f92e4d/snakeyaml-1.21.jar:/home/breandan/.gradle/caches/modules-2/files-2.1/com.google.code.gson/gson/2.8.6/9180733b7df8542621dc12e21e87557e8c99b8cb/gson-2.8.6.jar:/home/breandan/.gradle/caches/modules-2/files-2.1/net.java.dev.jna/jna/5.3.0/4654d1da02e4173ba7b64f7166378847db55448a/jna-5.3.0.jar:/home/breandan/.gradle/caches/modules-2/files-2.1/org.apache.commons/commons-compress/1.20/b8df472b31e1f17c232d2ad78ceb1c84e00c641b/commons-compress-1.20.jar:/home/breandan/.gradle/caches/modules-2/files-2.1/javax.inject/javax.inject/1/6975da39a7040257bd51d21a231b76c915872d38/javax.inject-1.jar:/home/breandan/.gradle/caches/modules-2/files-2.1/aopalliance/aopalliance/1.0/235ba8b489512805ac13a8f9ea77a1ca5ebe3e8/aopalliance-1.0.jar:/home/breandan/.gradle/caches/modules-2/files-2.1/com.google.guava/guava/19.0/6ce200f6b23222af3d8abb6b6459e6c44f4bb0e9/guava-19.0.jar:/home/breandan/IdeaProjects/djl/tensorflow/tensorflow-api/build/libs/tensorflow-api-0.11.0-SNAPSHOT.jar:/home/breandan/.gradle/caches/modules-2/files-2.1/org.bytedeco/javacpp/1.5.4/fb4e5f46cefbb6674c324bb28babc9fcf3caa642/javacpp-1.5.4.jar:/home/breandan/.gradle/caches/modules-2/files-2.1/org.tensorflow/tensorflow-core-api/0.3.0/97661aeface242b08853344459dd10a1c00cea57/tensorflow-core-api-0.3.0.jar:/home/breandan/.gradle/caches/modules-2/files-2.1/com.google.protobuf/protobuf-java/3.8.0/b5f93103d113540bb848fe9ce4e6819b1f39ee49/protobuf-java-3.8.0.jar:/home/breandan/.gradle/caches/modules-2/files-2.1/org.tensorflow/ndarray/0.3.0/7142502f27923196c1683b1a05eb369903e2e7a9/ndarray-0.3.0.jar
java.vm.vendor: AdoptOpenJDK
sun.arch.data.model: 64
user.variant: 
java.vendor.url: https://adoptopenjdk.net/
user.timezone: America/Toronto
os.name: Linux
java.vm.specification.version: 11
sun.java.launcher: SUN_STANDARD
user.country: US
sun.boot.library.path: /usr/lib/jvm/jdk-11.0.9+11/lib:/usr/lib/jvm/jdk-11.0.9+11/lib
sun.java.command: ai.djl.integration.util.DebugEnvironment
jdk.debug: release
sun.cpu.endian: little
user.home: /home/breandan
org.gradle.appname: gradlew
user.language: en
java.specification.vendor: Oracle Corporation
java.version.date: 2020-10-20
java.home: /usr/lib/jvm/jdk-11.0.9+11
ai.djl.logging.level: debug
org.gradle.internal.http.connectionTimeout: 60000
file.separator: /
java.vm.compressedOopsMode: Zero based
line.separator: 

java.specification.name: Java Platform API Specification
java.vm.specification.vendor: Oracle Corporation
java.awt.graphicsenv: sun.awt.X11GraphicsEnvironment
org.bytedeco.javacpp.platform.preloadpath: /home/breandan/.djl.ai/tensorflow/2.4.1-SNAPSHOT-20210317-cpu-linux-x86_64
sun.management.compiler: HotSpot 64-Bit Tiered Compilers
java.runtime.version: 11.0.9+11
user.name: breandan
path.separator: :
os.version: 4.15.0-132-generic
java.runtime.name: OpenJDK Runtime Environment
file.encoding: UTF-8
org.bytedeco.javacpp.maxPhysicalBytes: 0
jnidispatch.path: /home/breandan/.cache/JNA/temp/jna4749751000032380366.tmp
java.vm.name: OpenJDK 64-Bit Server VM
jna.platform.library.path: /usr/lib/x86_64-linux-gnu:/lib/x86_64-linux-gnu:/lib64:/usr/lib:/lib:/usr/lib/i386-linux-gnu:/lib/i386-linux-gnu:/lib32:/usr/lib32:/usr/lib/i386-linux-gnu/sse2:/usr/local/cuda-11.0/targets/x86_64-linux/lib:/usr/local/cuda/targets/x86_64-linux/lib:/usr/lib/x86_64-linux-gnu/libfakeroot
java.vendor.version: AdoptOpenJDK
jna.loaded: true
java.vendor.url.bug: https://github.com/AdoptOpenJDK/openjdk-support/issues
java.io.tmpdir: /tmp
org.gradle.internal.http.socketTimeout: 120000
java.version: 11.0.9
user.dir: /home/breandan/IdeaProjects/djl/integration
os.arch: amd64
java.vm.specification.name: Java Virtual Machine Specification
java.awt.printerjob: sun.print.PSPrinterJob
sun.os.patch.level: unknown
java.library.path: /usr/local/cuda-10.2/lib64:/usr/java/packages/lib:/usr/lib64:/lib64:/lib:/usr/lib
java.vm.info: mixed mode
java.vendor: AdoptOpenJDK
java.vm.version: 11.0.9+11
sun.io.unicode.encoding: UnicodeLittle
library.jansi.path: /home/breandan/.gradle/native/jansi/1.18/linux64
java.class.version: 55.0
org.gradle.internal.publish.checksums.insecure: true

--------- Environment Variables ---------
PATH: /home/breandan/anaconda3/condabin:/usr/local/cuda-10.2/bin:/usr/lib/jvm/jdk-11.0.9+11/bin:/home/breandan/.pyenv/plugins/pyenv-virtualenv/shims:/home/breandan/.pyenv/shims:/home/breandan/.pyenv/bin:/home/breandan/.cargo/bin:/home/breandan/.local/bin:/home/breandan/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/breandan/bin
PYENV_SHELL: bash
PM_PACKAGES_ROOT: /home/breandan/packman-repo
XAUTHORITY: /run/user/1000/gdm/Xauthority
XMODIFIERS: @im=ibus
XDG_DATA_DIRS: /usr/share/ubuntu:/home/breandan/.local/share/flatpak/exports/share/:/var/lib/flatpak/exports/share/:/usr/local/share/:/usr/share/:/var/lib/snapd/desktop
GDMSESSION: ubuntu
TEXTDOMAINDIR: /usr/share/locale/
GTK_IM_MODULE: ibus
CONDA_PYTHON_EXE: /home/breandan/anaconda3/bin/python
DBUS_SESSION_BUS_ADDRESS: unix:path=/run/user/1000/bus
XDG_CURRENT_DESKTOP: ubuntu:GNOME
SSH_AGENT_PID: 3721
COLORTERM: truecolor
LD_LIBRARY_PATH: /usr/local/cuda-10.2/lib64
QT4_IM_MODULE: xim
SESSION_MANAGER: local/romulus:@/tmp/.ICE-unix/3643,unix/romulus:/tmp/.ICE-unix/3643
USERNAME: breandan
LOGNAME: breandan
PWD: /home/breandan/IdeaProjects/djl
IM_CONFIG_PHASE: 2
PYENV_VIRTUALENV_INIT: 1
GJS_DEBUG_TOPICS: JS ERROR;JS LOG
SHELL: /bin/bash
LESSOPEN: | /usr/bin/lesspipe %s
OLDPWD: /home/breandan/IdeaProjects/djl
GNOME_DESKTOP_SESSION_ID: this-is-deprecated
GNOME_TERMINAL_SCREEN: /org/gnome/Terminal/screen/57a34145_dbd0_4bdf_bb7b_74a4b6e3a849
GTK_MODULES: gail:atk-bridge
CLUTTER_IM_MODULE: xim
LC_ALL: en_US.UTF-8
TEXTDOMAIN: im-config
XDG_SESSION_DESKTOP: ubuntu
LS_COLORS: rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.Z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:
SHLVL: 1
LESSCLOSE: /usr/bin/lesspipe %s %s
QT_IM_MODULE: ibus
CONDA_EXE: /home/breandan/anaconda3/bin/conda
TERM: xterm-256color
XDG_CONFIG_DIRS: /etc/xdg/xdg-ubuntu:/etc/xdg
GNOME_TERMINAL_SERVICE: :1.3215
LANG: en_US.UTF-8
XDG_SESSION_ID: 2
XDG_SESSION_TYPE: x11
DISPLAY: :1
DTSHELL_COMMANDS: /home/breandan/CLionProjects/duckietown-shell-commands
_CE_M: 
CONDA_SHLVL: 0
_: ./gradlew
GPG_AGENT_INFO: /run/user/1000/gnupg/S.gpg-agent:0:1
DESKTOP_SESSION: ubuntu
USER: breandan
XDG_MENU_PREFIX: gnome-
VTE_VERSION: 5202
QT_ACCESSIBILITY: 1
WINDOWPATH: 2
GJS_DEBUG_OUTPUT: stderr
SSH_AUTH_SOCK: /run/user/1000/keyring/ssh
XDG_SEAT: seat0
_CE_CONDA: 
GNOME_SHELL_SESSION_MODE: ubuntu
XDG_VTNR: 2
XDG_RUNTIME_DIR: /run/user/1000
HOME: /home/breandan

-------------- Directories --------------
temp directory: /tmp
Engine cache directory: /home/breandan/.djl.ai

------------------ CUDA -----------------
GPU Count: 1
Default Device: gpu(0)
CUDA: 102
ARCH: 52
GPU(0) memory used: 2685534208 bytes

----------------- Engines ---------------
Default Engine: MXNet
[DEBUG] - Using cache dir: /home/breandan/.djl.ai/pytorch
PyTorch:1.7.1, capabilities: [
        CUDA,
        CUDNN,
        OPENMP,
        MKL,
        MKLDNN,
]
PyTorch Library: /home/breandan/.djl.ai/pytorch/1.7.1-cu102-linux-x86_64/0.11.0-SNAPSHOT-cu102-libdjl_torch.so
[DEBUG] - Using cache dir: /home/breandan/.djl.ai/mxnet
MXNet:1.7.0, capabilities: [
        CUDNN,
        SIGNAL_HANDLER,
        LAPACK,
        CPU_SSE2,
        CPU_SSE3,
        OPENCV,
        CUDA,
        CPU_SSE,
        CPU_AVX,
        F16C,
        BLAS_OPEN,
        NCCL,
        CPU_SSE4_2,
        DIST_KVSTORE,
        CUDA_RTC,
        CPU_SSE4_1,
        OPENMP,
        MKLDNN,
]
MXNet Library: /home/breandan/.djl.ai/mxnet/1.7.0-backport-cu102mkl-linux-x86_64/libmxnet.so
[DEBUG] - Using cache dir: /home/breandan/.djl.ai/tensorflow
[WARN ] - No matching CUDA flavor for linux found: cpu/sm_52, fallback to CPU.
TensorFlow:2.4.1, capabilities: [
        MKL,
]
TensorFlow Library: /home/breandan/.djl.ai/tensorflow/2.4.1-SNAPSHOT-20210317-cpu-linux-x86_64/libjnitensorflow.so

--------------- Hardware --------------
Available processors (cores): 8
Byte Order: LITTLE_ENDIAN
Free memory (bytes): 733290528
Maximum memory (bytes): 16852713472
Total memory available to JVM (bytes): 1056964608
Heap committed: 1056964608
Heap nonCommitted: 41140224
GCC: 
gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.7.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 2m 10s
39 actionable tasks: 19 executed, 20 up-to-date
```

--- a/api/src/main/java/ai/djl/ndarray/BaseNDManager.java
+++ b/api/src/main/java/ai/djl/ndarray/BaseNDManager.java
@@ -15,7 +15,6 @@ package ai.djl.ndarray;
 import ai.djl.Device;
 import ai.djl.ndarray.types.DataType;
 import ai.djl.ndarray.types.Shape;
-import ai.djl.util.Pair;
 import ai.djl.util.PairList;
 import java.nio.Buffer;
 import java.nio.file.Path;
@@ -35,7 +34,7 @@ public abstract class BaseNDManager implements NDManager {
     protected String name;
     protected Device device;
     protected ConcurrentHashMap<String, AutoCloseable> resources;
-    protected ConcurrentHashMap<String, Pair<NDResource, NDManager>> tempResources;
+    protected ConcurrentHashMap<String, TempResource> tempResources;
     protected AtomicBoolean closed = new AtomicBoolean(false);
 
     protected BaseNDManager(NDManager parent, Device device) {
@@ -210,7 +209,22 @@ public abstract class BaseNDManager implements NDManager {
         if (closed.get()) {
             throw new IllegalStateException("NDManager has been closed already.");
         }
-        resources.put(resourceId, resource);
+        tempResources.compute(
+                resourceId,
+                (key, tempResource) -> {
+                    if (tempResource != null) {
+                        // This state occurs when this manager (manA) tempAttaches a resource that
+                        // is later
+                        // tempAttached to another manager (manB)
+                        // When manB is closed, it will use attach to return the resource to this
+                        // (manA)
+                        // In that case, it should stay as a tempResource in this (manA)
+                        tempResource.detached = false;
+                    } else {
+                        resources.put(resourceId, resource);
+                    }
+                    return tempResource;
+                });
     }
 
     /** {@inheritDoc} */
@@ -220,7 +234,7 @@ public abstract class BaseNDManager implements NDManager {
         if (closed.get()) {
             throw new IllegalStateException("NDManager has been closed already.");
         }
-        tempResources.put(resourceId, new Pair<>(resource, originalManager));
+        tempResources.put(resourceId, new TempResource(resource, originalManager));
     }
 
     /** {@inheritDoc} */
@@ -230,6 +244,12 @@ public abstract class BaseNDManager implements NDManager {
             // This may happen in the middle of BaseNDManager.close()
             return;
         }
+        tempResources.computeIfPresent(
+                resourceId,
+                (key, tempResource) -> {
+                    tempResource.detached = true;
+                    return tempResource;
+                });
         resources.remove(resourceId);
     }
 
@@ -257,15 +277,18 @@ public abstract class BaseNDManager implements NDManager {
                     logger.error("Resource close failed.", e);
                 }
             }
-            for (Pair<NDResource, NDManager> resource : tempResources.values()) {
+            for (TempResource resource : tempResources.values()) {
                 try {
-                    resource.getKey().attach(resource.getValue());
+                    if (!resource.detached) {
+                        resource.resource.attach(resource.manager);
+                    }
                 } catch (Exception e) {
                     logger.error("Temporary resource return failed.", e);
                 }
             }
             parent.detachInternal(uid);
             resources.clear();
+            tempResources.clear();
         }
     }
 
@@ -289,6 +312,18 @@ public abstract class BaseNDManager implements NDManager {
             if (c instanceof BaseNDManager) {
                 ((BaseNDManager) c).debugDump(level + 1);
             }
+        }
+    }
+
+    protected static final class TempResource {
+        private NDResource resource;
+        private NDManager manager;
+        private boolean detached;
+
+        public TempResource(NDResource resource, NDManager manager) {
+            this.resource = resource;
+            this.manager = manager;
+            this.detached = false;
         }
     }
 }

--- a/api/src/main/java/ai/djl/nn/transformer/PointwiseFeedForwardBlock.java
+++ b/api/src/main/java/ai/djl/nn/transformer/PointwiseFeedForwardBlock.java
@@ -13,24 +13,14 @@
 package ai.djl.nn.transformer;
 
 import ai.djl.ndarray.NDList;
-import ai.djl.ndarray.NDManager;
-import ai.djl.ndarray.types.DataType;
-import ai.djl.ndarray.types.Shape;
-import ai.djl.nn.AbstractBlock;
-import ai.djl.nn.Block;
 import ai.djl.nn.LambdaBlock;
+import ai.djl.nn.SequentialBlock;
 import ai.djl.nn.core.Linear;
-import ai.djl.training.ParameterStore;
-import ai.djl.util.Pair;
-import ai.djl.util.PairList;
-import java.util.Collections;
 import java.util.List;
 import java.util.function.Function;
 
 /** Fully connected Feed-Forward network, only applied to the last dimension of the input. */
-public class PointwiseFeedForwardBlock extends AbstractBlock {
-
-    private static final byte VERSION = 1;
+public class PointwiseFeedForwardBlock extends SequentialBlock {
 
     /**
      * Creates a pointwise feed-forward block.
@@ -44,59 +34,12 @@ public class PointwiseFeedForwardBlock extends AbstractBlock {
             List<Integer> hiddenSizes,
             int outputSize,
             Function<NDList, NDList> activationFunction) {
-        super(VERSION);
         // add hidden layers with activation
-        int count = 0;
         for (int hiddenSize : hiddenSizes) {
-            addChildBlock(
-                    "linear_" + count, Linear.builder().optBias(true).setUnits(hiddenSize).build());
-            addChildBlock("activation_" + count, new LambdaBlock(activationFunction));
-            ++count;
+            add(Linear.builder().optBias(true).setUnits(hiddenSize).build());
+            add(new LambdaBlock(activationFunction));
         }
         // add output layer without activation
-        addChildBlock("output_layer", Linear.builder().optBias(true).setUnits(outputSize).build());
-    }
-
-    /** {@inheritDoc} */
-    @Override
-    public Shape[] getOutputShapes(Shape[] inputShapes) {
-        for (Block child : children.values()) {
-            inputShapes = child.getOutputShapes(inputShapes);
-        }
-        return inputShapes;
-    }
-
-    /** {@inheritDoc} */
-    @Override
-    public void initializeChildBlocks(NDManager manager, DataType dataType, Shape... inputShapes) {
-        inputNames = Collections.singletonList("input");
-        if (inputShapes.length != 1) {
-            throw new IllegalArgumentException(
-                    "Pointwise feed forward blocks can only have one input.");
-        }
-        // Now that we know the input shape, we can determine the reshape necessary
-        // to shape the input and re-shape the output
-        Shape inputShape = inputShapes[0];
-        if (inputShape.dimension() < 2) {
-            throw new IllegalArgumentException(
-                    "Pointwise feed forward blocks need an input of at least dimension 2.");
-        }
-        Shape lastShape = inputShape;
-        for (Block child : children.values()) {
-            child.initialize(manager, dataType, lastShape);
-            lastShape = getOutputShapes(new Shape[] {lastShape})[0];
-        }
-    }
-
-    /** {@inheritDoc} */
-    @Override
-    protected NDList forwardInternal(
-            ParameterStore ps, NDList inputs, boolean training, PairList<String, Object> params) {
-        // go through all layers
-        NDList layerResult = inputs;
-        for (Pair<String, Block> child : getChildren()) {
-            layerResult = child.getValue().forward(ps, layerResult, training);
-        }
-        return layerResult;
+        add(Linear.builder().optBias(true).setUnits(outputSize).build());
     }
 }

--- a/mxnet/mxnet-engine/src/main/java/ai/djl/mxnet/engine/MxNDManager.java
+++ b/mxnet/mxnet-engine/src/main/java/ai/djl/mxnet/engine/MxNDManager.java
@@ -20,6 +20,7 @@ import ai.djl.ndarray.BaseNDManager;
 import ai.djl.ndarray.NDArray;
 import ai.djl.ndarray.NDList;
 import ai.djl.ndarray.NDManager;
+import ai.djl.ndarray.NDResource;
 import ai.djl.ndarray.types.DataType;
 import ai.djl.ndarray.types.Shape;
 import ai.djl.ndarray.types.SparseFormat;
@@ -386,6 +387,11 @@ public class MxNDManager extends BaseNDManager {
         /** {@inheritDoc} */
         @Override
         public void attachInternal(String resourceId, AutoCloseable resource) {}
+
+        /** {@inheritDoc} */
+        @Override
+        public void tempAttachInternal(
+                NDManager originalManager, String resourceId, NDResource resource) {}
 
         /** {@inheritDoc} */
         @Override

